### PR TITLE
Feat/autodiff/checkpoint ops

### PR DESCRIPTION
**Pull request by louisfd**
Finally, a real PR for autodiff checkpointing. 
There is now a way to do no checkpointing by considering everything to be compute bound. It is set by default, so this should be able to merge without impacting anything.

### CODE REVIEW

File: crates/burn-autodiff/src/ops/base.rs [hunk 1]
```diff
@@ -107,20 +278,30 @@ where
...
         let parents = self.nodes.map(|node| node.clone_if_require_grad());
         let ops = Ops::new(parents, output.node.clone(), state);
 
         output.register_step(OpsStep::new(ops, self.backward))
     }
+
+    /// Checkpoints the tensor
+    pub fn checkpoint<const D2: usize>(&mut self, tensor: &AutodiffTensor<B, D2>) -> NodeID {
+        self.checkpointer_builder
+            .checkpoint(tensor, ActionType::Main);
```
Comment (nathanielsimard) [inline crates/burn-autodiff/src/ops/base.rs:293]: Main => Explicit maybe?

File: crates/burn-autodiff/src/checkpoint/builder.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        backward_states_map: &mut HashMap<NodeID, State>,
+        retro_forward_map: &mut HashMap<NodeID, Arc<dyn RetroForward>>,
+        n_required_map: HashMap<NodeID, usize>,
+    ) {
+        // We do not loop over checkpointing actions anymore because they can contain
+        // duplicates or miss some that are in backup
+        for (node_id, n_required) in n_required_map {
+            // We find the checkpointing action for node_id. It's likely in checkpointing_actions
+            // so we check there first, otherwise it will be in backup.
+            // Technically it can be there several times but can never be of both types, so we can assume the first we find is fine
```
[the remark this answers is not in the document]
Comment (louisfd) [reply crates/burn-autodiff/src/checkpoint/builder.rs:200]: Actually, although we iterate on the n_required_map, we find the checkpointing action in the checkpointing actions Vec, using the id in the map. The vec can have duplicates. I'll rework the text so it's clearer

File: crates/burn-autodiff/src/checkpoint/state.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        let state = self.map.remove(node_id).unwrap();
+        let remaining_n_required = state.n_required() - 1;
+
+        // Downcast the state to whatever it is supposed to be
+        // If still needed after giving ownership, we copy it back to the hashmap
+        if remaining_n_required > 0 {
+            let new_stored_state = match state {
+                State::Recompute { n_required: _ } => State::Recompute {
+                    n_required: remaining_n_required,
+                },
```
[the remark this answers is not in the document]
Comment (louisfd) [reply crates/burn-autodiff/src/checkpoint/state.rs:86]: good catch, indeed i can put unreachable and everything passes

File: crates/burn-autodiff/src/ops/base.rs [hunk 4]
```diff
@@ -1,37 +1,198 @@
...
+                self.backward,
+                self.compute_property,
+                self.checkpointer_builder,
+            )),
+        }
+    }
+}
+
+/// Duplicated for Init because we can choose to skip compute property chosen (defaults to ambiguous)
+impl<BO, B, C, const D: usize, const N: usize> OpsPrep<BO, B, (), C, D, N, Init>
```
[the remark this answers is not in the document]
Comment (louisfd) [reply crates/burn-autodiff/src/ops/base.rs:195]: Indeed. Ok I'll remove it now that all ops have chosen between memory and compute

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**